rasterio/_warp.pyx

```diff
@@ -348,150 +352,138 @@ def _reproject(
+        psWOptions.padfDstNoDataImag[i] = 0.0
+    warp_extras = _gdal.CSLSetNameValue(
+        warp_extras, "INIT_DEST", "NO_DATA")
+
+    # Important: set back into struct or values set above are lost
+    # This is because CSLSetNameValue returns a new list each time
+    psWOptions.papszWarpOptions = warp_extras
+
+    psWOptions.pfnTransformer = _gdal.GDALGenImgProjTransform
+    psWOptions.pTransformerArg = hTransformArg
+    psWOptions.hSrcDS = hdsin
+    psWOptions.hDstDS = hdsout
+    psWOptions.nBandCount = src_count
+    psWOptions.panSrcBands = <int *>_gdal.CPLMalloc(src_count*sizeof(int))
+    psWOptions.panDstBands = <int *>_gdal.CPLMalloc(src_count*sizeof(int))
+    if isinstance(source, tuple):
+        psWOptions.panSrcBands[0] = source.bidx
+    else:
+        for i in range(src_count):
+            psWOptions.panSrcBands[i] = i+1
+    if isinstance(destination, tuple):
+        psWOptions.panDstBands[0] = destination.bidx
+    else:
+        for i in range(src_count):
+            psWOptions.panDstBands[i] = i+1
+    log.debug("Set transformer options")
 
-        eErr = oWarper.Initialize(psWOptions)
-        if eErr == 0:
+    # TODO: alpha band.
 
-            log.debug("Destination shape: %r", destination.shape)
+    if oWarper.Initialize(psWOptions):
+        raise RuntimeError("Failed to initialize warper.")
 
-            rows, cols = destination.shape[-2:]
-            log.debug(
-                "Chunk and warp window: %d, %d, %d, %d",
-                0, 0, cols, rows)
-            with nogil:
-                eErr = oWarper.ChunkAndWarpMulti(0, 0, cols, rows)
-            log.debug("Chunked and warped: %d", retval)
-    
-    except Exception:
-        log.exception(
-            "Caught exception in warping. Source not reprojected.")
-        raise
-    
     else:
-        reprojected = True
-
-    finally:
-        if hTransformArg != NULL:
-            _gdal.GDALDestroyGenImgProjTransformer(hTransformArg)
-        #if maxerror > 0.0:
-        #    _gdal.GDALDestroyApproxTransformer(psWOptions.pTransformerArg)
-        if psWOptions != NULL:
-            _gdal.GDALDestroyWarpOptions(psWOptions)
-        if dtypes.is_ndarray(source):
-            if hdsin != NULL:
-                _gdal.GDALClose(hdsin)
+        rows, cols = destination.shape[-2:]
+        log.debug(
+            "Chunk and warp window: %d, %d, %d, %d.",
+            0, 0, cols, rows)
+
+        if num_threads > 1:
+            err_code = oWarper.ChunkAndWarpMulti(0, 0, cols, rows)
+        else:
+            err_code = oWarper.ChunkAndWarpImage(0, 0, cols, rows)
+
+        log.debug("Chunked and warped: %d", err_code)
+
+        reprojected = not(err_code)
+
+    if hTransformArg != NULL:
+        _gdal.GDALDestroyGenImgProjTransformer(hTransformArg)
+    if psWOptions != NULL:
+        _gdal.GDALDestroyWarpOptions(psWOptions)
+    if dtypes.is_ndarray(source):
+        if hdsin != NULL:
+            _gdal.GDALClose(hdsin)
```

Comment:
The churn here is mostly to do with removing the big and unnecessary `try/except` block.